### PR TITLE
Feature branch

### PR DESCRIPTION
**Pull request by AbdelrhmanAmin**
# ** !!!COULDN'T FIX LINTERS PLEASE TELL ME HOW TO FIX IT!!!! **
# Advanced-Building-Blocks---Enumerables

In this project, I've built my own version of the built-in Ruby’s enumerable methods:

- Each;
- Each with index;
- Select;
- All;
- Any;
- None;
- Count;
- Map;
- Inject.

### CODE REVIEW

**Comment by AbdelrhmanAmin** [comment]
> _No description provided._

File: index.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+          else
+            if !opt.nil? && flag == true
+              opt.to_proc(sum, elem)
+            else
+              yield(sum, elem)
+                  end
+          end
+  end
+  sum
+end
```
Comment (kitpao) [inline index.rb]: - [x] be careful with your indentation so that you have less trouble finding errors in your code

File: README.md [hunk 2]
```diff
@@ -1 +1,42 @@
...
+## Built With
+
+- Ruby
+
+## Getting Started
+
+- open CMD
+- git clone https://github.com/AbdelrhmanAmin/Advanced-Building-Blocks---Enumerables.git
+- **Make sure Ruby is installed on your computer**
+- in terminal type: ruby bubble_sort.rb
```
Comment (kitpao) [inline README.md]: - [x] Please update the instructions for this project (instead of bubble_sort.rb)

File: Gemfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+source 'https://rubygems.org'
+
+git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
+
+gem 'rubocop', '~>0.81.0'
+
+# gem "rails"
```
Comment (kitpao) [inline Gemfile]: - [x] Please delete commented code
image

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @AbdelrhmanAmin 

- [x] The reason Rubocop is complaining is that you have an extra `end` in line 106. Please check all your indentation, it is causing issues in your logic. Remember you have to use 2 spaces of indentation when you use a loop or an if and leave everything in the same level in the same indentation line. Finally, run Rubocop again, you will have different errors, I will leave you an example of good indentation:
image


- [x] The second error you have right now in Rubocop is in the test.rb file, this file is not required, you can add it to gitignore. But if you want to fix it, you just have to delete the parenthesis -> `puts 1..5.my_map(&my_proc)`

### Methods:
  **Tip:** You can see the complete [description of all the methods here](https://ruby-doc.org/core-2.6.5/Enumerable.html), please check it out as a guide if you feel lost

#### #my_each
- [x] when a block is given when self is a range returns the range itself after calling the given block once for each element in self -> This means you have to return the same input (return at the end self, you are returning the modified input converted into an array, take the case of ranges)

#### #my_each_with_index
- [x] when a block is given when self is an array returns the array itself after calling the given block once for each element in self -> This means you have to return the same input (return at the end self, you are returning the modified input converted into an array, take the case of ranges)

#### #my_all?
  - [x] Please notice the difference between expected and result:
  ```
  p [1,2,3].all?                 # correct answer: false
  p [1,2,3].my_all?(3)      # your method returns true
  ```

#### #my_any?
  - [x] `['e', 'dd'].my_any?(/d/)` should be true because one of the elements matches with /d/ regexp
  - [x] `[1,2,3].my_any?(3)` should be true because one of the elements matches with 3

#### #my_none?
  - [x] `['e', 'dd'].my_none?(/d/)` should be false because one of the elements matches with /d/ regexp
  - [x] `[1,2,3].my_none?(3)` should be false because one of the elements matches with 3

#### #my_count
  - [x] Please notice this method should not have "?" in the name

#### #my_map
  - [x] This method should return an enumerator if no block is given

#### #my_inject
   - [x] Your method should also be able to multiply elements without a block, `[2,3].my_inject(20, :*)` should return 120 (20 * 2 * 3)
  - [x]  `[2,3,4].my_inject(:+)` should return 9 -> 2+3+4=9. Your code gives an error if no block is given

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

File: README.md [hunk 4]
```diff
@@ -1 +1,42 @@
...
+## Built With
+
+- Ruby
+
+## Getting Started
+
+- open CMD
+- git clone https://github.com/AbdelrhmanAmin/Advanced-Building-Blocks---Enumerables.git
+- **Make sure Ruby is installed on your computer**
+- in terminal type: ruby index.rb
```
Comment (kitpao) [inline README.md]: - [x] **OPTIONAL** If you would like to format your terminal instructions as code, you can use backquotes
```
`like this`
```

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @AbdelrhmanAmin 

You progressed quite a lot 👏 Most tests are passing and you managed to fix issues with the linters (you can edit and take out that message from the pull request message if you want)

You still have test cases in 3 methods that are not passing, please check below 🔽 

#### #my_any?
- [x] Seems like your code is now not working with strings as arguments `["car","cat","dog"].my_any?("cat")` should be true

#### #my_none?
- [x] Same happens in my_none: `["car","cat","dog"].my_none?("cat")` should be false because there is a cat.
- [x] `p [1,2,2].my_none?(3)` should be true because there is no 3

#### #my_inject
  - [x] If there is no block or argument, your method should return a local jump error ([how to raise errors here](https://stackoverflow.com/questions/4325507/how-to-force-a-localjumperror-on-return-in-ruby))

That is all 😄 you are almost there 🏅 

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by kitpao — APPROVED** [review]
# 👋 Hi again @AbdelrhmanAmin 

You have done amazingly 🥳 All the tests are passing and 
### 👏 Your project has been approved 👏
### 🚀 You may merge to development and submit the project completition form 🚀

___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Pull request merged**